### PR TITLE
Features/task 3

### PR DESCRIPTION
**Pull request by Neytus**

### CODE REVIEW

File: TodoList/sources/TodoList.Api/Controllers/NodesController.cs [hunk 1]
```diff
@@ -5,40 +5,112 @@
...
+            }
+
+            return Created(_locationHelper.GetNodeLocation(newNode.Id), newNode);
         }
 
-        public async Task<IHttpActionResult> PutAsync(NodeModel model)
-            => Content(HttpStatusCode.Accepted, await _repository.UpdateAsync(model));
+        public async Task<IHttpActionResult> PutAsync(NodeModel node)
+        {
+            if (!ValidateNodeText(node.Text)) return BadRequest(ModelState);
```
Comment (PetrSvirak) [inline TodoList/sources/TodoList.Api/Controllers/NodesController.cs]: This will fail if there is no `node` model passed to the method (e.g., when sent `JSON`, is malformatted)

File: TodoList/sources/TodoList.Api/Controllers/NodesController.cs [hunk 2]
```diff
@@ -5,40 +5,112 @@
...
+            {
+                return BadRequest(exception.Message);
+            }
+
+            return Created(_locationHelper.GetNodeLocation(newNode.Id), newNode);
         }
 
-        public async Task<IHttpActionResult> PutAsync(NodeModel model)
-            => Content(HttpStatusCode.Accepted, await _repository.UpdateAsync(model));
+        public async Task<IHttpActionResult> PutAsync(NodeModel node)
```
Comment (PetrSvirak) [inline TodoList/sources/TodoList.Api/Controllers/NodesController.cs]: * I like how completely you implemented the `PUT` 👍 
* Consider re-adding `id` – I did not realize that omitting it would not be restful since you manipulate single node and not a collection of them. Do not forget that `id` in `URL` and `body` might eventually differ (which is wrong (unless you decided to support objects without id specified (then it would be OK for a node object to (must) have `Guid.Empty`)))
* You should not accept a node with either `Creation` or `LastUpdate` specified – this is one of the places where you would normally accept a different model and map it (yet using `AutoMapper` would really be overkill, so let's put this into a validation method dedicated for the `put` method)

File: TodoList/sources/TodoList.Api/Controllers/NodesController.cs [hunk 3]
```diff
@@ -5,40 +5,112 @@
...
             _locationHelper = locationHelper;
         }
 
         public async Task<IHttpActionResult> GetAsync()
             => Ok(await _repository.GetAllAsync());
 
         public async Task<IHttpActionResult> GetAsync(Guid id)
-            => Ok(await _repository.GetByIdAsync(id));
+        {
+            if (!ModelState.IsValid) return BadRequest(ModelState);
```
Comment (PetrSvirak) [inline TodoList/sources/TodoList.Api/Controllers/NodesController.cs]: I don't like these oneliners at all. Preferably wrap it in curly braces (read [this discussion](https://social.msdn.microsoft.com/Forums/vstudio/en-US/890c2266-5fab-486d-a900-4bb27e6fedca/omit-curly-braces-or-not?forum=csharpgeneral) for examples and reasons) – it enhances readability and prevents a misreading error – when formatting suggests a command is executed conditionally but it is actually executed every time: 
````
if(false)
    Console.WriteLine("Never");
    Console.WriteLine("But Always");
````

File: TodoList/sources/TodoList.Api/Controllers/NodesController.cs [hunk 4]
```diff
@@ -5,40 +5,112 @@
...
+
+            var returnedNode = await _repository.GetByIdAsync(id);
```
Comment (PetrSvirak) [inline TodoList/sources/TodoList.Api/Controllers/NodesController.cs]: Well, it is obviously `returned`, but let's give the variable a bit more meaning – such as `existingNode`?

File: TodoList/sources/TodoList.Api/Controllers/NodesController.cs [hunk 5]
```diff
@@ -5,40 +5,112 @@
...
+            if (!ValidatePostNodeModel(node)) return BadRequest(ModelState);
+            if (!ModelState.IsValid) return BadRequest(ModelState);
+
+            NodeModel newNode;
+
+            try
+            {
+                newNode = await _createNodeService.CreateNodeAsync(node);
+            }
+            catch (Exception exception)
```
Comment (PetrSvirak) [inline TodoList/sources/TodoList.Api/Controllers/NodesController.cs]: This is not a good way of handling exceptions – it is repetitive and makes each method unnecessarily longer. You should either introduce a middleware to rewrite all exceptions to a `BadRequest` or rewrite the controller's code to reacts to specific exceptions. In our case, we will not handle any exceptions caused by Mongo (wrong connection and so forth) and write service layer in such way that other than internal server errors will won't be thrown. And we will not introduce middleware to handle these errors, so it will naturally end up with `Internal server error` (which is OK as far there is no exception thrown by our code directly).

File: TodoList/sources/TodoList.Api/Controllers/NodesController.cs [hunk 6]
```diff
@@ -5,40 +5,112 @@
...
+            if (!ModelState.IsValid) return BadRequest(ModelState);
+
+            if (await _repository.GetByIdAsync(id) == null) return NotFound();
+
             await _repository.DeleteAsync(id);
+
             return Ok();
         }
+
+        private bool ValidateNodeText(string text)
```
Comment (PetrSvirak) [inline TodoList/sources/TodoList.Api/Controllers/NodesController.cs]: Do not return `boolean` from these methods, check `ModelState.IsValid` in the `action`s and return `BadRequest` based on it. You can alter `ModelState` by calling its `AddError` method.

File: TodoList/sources/TodoList.Contracts/Services/IUpdateNodeService.cs [hunk 7]
```diff
@@ -0,0 +1,10 @@
+﻿using System.Threading.Tasks;
+using TodoList.Contracts.Models;
+
+namespace TodoList.Contracts.Services
+{
+    public interface IUpdateNodeService
```
Comment (PetrSvirak) [inline TodoList/sources/TodoList.Contracts/Services/IUpdateNodeService.cs:7]: This interface actually provides great room for optimization. You actually know that updating a node precedes a search for existing one in the DB. If you do not optimize your update service, two requests to the database might actually be filed. You should use your service to cache the node after calling a method called on it that checks if it exists and then re-use the node when actually merging the nodes together before updating it (within the only current method of the interface)…

File: TodoList/sources/TodoList.Repository/Helpers/ConnectionHelper.cs [hunk 8]
```diff
@@ -0,0 +1,12 @@
+﻿using System.Configuration;
+
+namespace TodoList.Repository.Helpers
+{
+    internal static class ConnectionHelper
+    {
+        internal static string GetDbConnection()
```
Comment (PetrSvirak) [inline TodoList/sources/TodoList.Repository/Helpers/ConnectionHelper.cs]: * Make it a lambda, definitely
* Create a contract for it, the reason this class works is that you always run in from `WebAPI`. It would be much better to leverage `inversion of control` (or `dependency injection` more precisely) to define an interface, much like [`options pattern`](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration?tabs=basicconfiguration#using-options-and-configuration-objects) in .Net Core. This contract will have its implementation in `API` layer while `repository` will require an instance fulfilling the contract in its constructor.
* Move the call of `ConfigurationManager` to constructor – it is OK to cache it at least per request.

File: TodoList/sources/TodoList.Repository/NodesRepository.cs [hunk 9]
```diff
@@ -1,36 +1,39 @@
...
-            });
+            => (await _dbCollection.FindAsync(FilterDefinition<NodeModel>.Empty)).ToEnumerable();
 
         public async Task<NodeModel> GetByIdAsync(Guid id)
-            => await Task.FromResult(FirstModel);
+            => await _dbCollection.Find(node => node.Id == id).FirstOrDefaultAsync();
 
-        public async Task<NodeModel> AddAsync(NodeModel model)
-            => await Task.FromResult(SecondModel);
+        public async Task AddAsync(NodeModel model)
```
Comment (PetrSvirak) [inline TodoList/sources/TodoList.Repository/NodesRepository.cs]: Revert the changes in `INodesRepository` contract and find a way to implement it with mongo driver 😉

File: TodoList/sources/TodoList.Services/Dependency/RegisterTypes.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+namespace TodoList.Services.Dependency
+{
+    public class RegisterTypes : IBootstrapper
+    {
+        public IUnityContainer RegisterType(IUnityContainer container)
+        {
+            container.RegisterType<IGenerateIdService, GenerateIdService>(new ContainerControlledLifetimeManager());
+            container.RegisterType<ICreateNodeService, CreateNodeService>(new ContainerControlledLifetimeManager());
+            container.RegisterType<ICurrentTimeService, CurrentTimeService>(new ContainerControlledLifetimeManager());
+            container.RegisterType<IUpdateNodeService, UpdateNodeService>(new ContainerControlledLifetimeManager());
```
Comment (PetrSvirak) [inline TodoList/sources/TodoList.Services/Dependency/RegisterTypes.cs]: You actually need `HierarchicalLifetimeManager` – read on why [here](https://msdn.microsoft.com/en-us/library/ff660872(v=pandp.20).aspx)

File: TodoList/sources/TodoList.Services/UpdateNodeService.cs [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+        {
+            _repository = repository;
+            _timeService = timeService;
+        }
+
+        public async Task<NodeModel> UpdateNodeAsync(NodeModel node)
+        {
+            var currentTime = _timeService.GetCurrentTime();
+
+            var chosenNode = await _repository.GetByIdAsync(node.Id);
```
Comment (PetrSvirak) [inline TodoList/sources/TodoList.Services/UpdateNodeService.cs]: * Reminds me of Neo :rofl: 
* Try to be consistent in naming - I pointed out that you could use `existing` prefix in the `Controller`; whatever you decide to use, consider using it here as well
* Rename the `node` as well to express value's purpose or origin

File: TodoList/sources/TodoList.Repository/Dependency/RegisterTypes.cs [hunk 12]
```diff
@@ -1,12 +1,13 @@
...
 using TodoList.Contracts.Repository;
+using TodoList.Repository.Helpers;
 
 namespace TodoList.Repository.Dependency
 {
     public class RegisterTypes : IBootstrapper
     {
         public IUnityContainer RegisterType(IUnityContainer container) 
-            => container.RegisterType<INodesRepository, NodesRepository>(new HierarchicalLifetimeManager());
+            => container.RegisterType<INodesRepository, NodesRepository>(new InjectionConstructor(ConnectionHelper.GetDbConnection()));
```
Comment (PetrSvirak) [inline TodoList/sources/TodoList.Repository/Dependency/RegisterTypes.cs]: You do not specify a lifetime manager here (making it [transient](https://msdn.microsoft.com/en-us/library/microsoft.practices.unity.transientlifetimemanager.aspx)) and you really should. Read through `Mongo` driver's documentation and select the correct manager, please.

**Review by PetrSvirak — CHANGES_REQUESTED** [review]
I really like your solution as a whole. There certain things that need a bit more tinkering, but in general, it is all good, great work.

* I did not review tests much (since I want to wait with it for the other code's final form), however, I noticed you have some tests for `null` cases missing.
* In `Services` assembly:
 * move the Node services to a standalone folder.
 * rename `Helpers` folder to something more fitting - these can, for example, be called static wrappers (since they wrap static calls, …) 
* Do not call a class implementing `IBootstrapper` interface `RegisterTypes` → `Bootstrapper` would be more fitting (be careful about git history)
* Do not use `Helper` suffix (or any other general suffix) when it can be omitted

File: TodoList/sources/TodoList.Api/Controllers/NodesController.cs [hunk 13]
```diff
@@ -5,40 +5,161 @@
...
+            }
+
+            var newNode = await _createNodeService.CreateNodeAsync(node);
+
+            return Created(_locator.GetNodeLocation(newNode.Id), newNode);
         }
 
-        public async Task<IHttpActionResult> PutAsync(NodeModel model)
-            => Content(HttpStatusCode.Accepted, await _repository.UpdateAsync(model));
+        public async Task<IHttpActionResult> PutAsync([FromBody] NodeModel node)
```
Comment (vit-svoboda) [inline TodoList/sources/TodoList.Api/Controllers/NodesController.cs]: In REST, the PUT action should use the id from the route. Note that then you'll have 2 possibly different identifiers for the same node and you need to deal with that somehow. Either you just ignore the one from the body, or you introduce a view-model for PUT body. It's up to you, the former is easier.

File: TodoList/sources/TodoList.Contracts/TodoList.Contracts.csproj [hunk 14]
```diff
@@ -42,7 +42,14 @@
     <Reference Include="Microsoft.Practices.Unity.RegistrationByConvention, Version=4.0.0.0, Culture=neutral, PublicKeyToken=6d32ff45e0ccc69f, processorArchitecture=MSIL">
       <HintPath>..\..\packages\Unity.4.0.1\lib\net45\Microsoft.Practices.Unity.RegistrationByConvention.dll</HintPath>
     </Reference>
+    <Reference Include="MongoDB.Bson, Version=2.4.4.0, Culture=neutral, processorArchitecture=MSIL">
+      <HintPath>..\..\packages\MongoDB.Bson.2.4.4\lib\net45\MongoDB.Bson.dll</HintPath>
+    </Reference>
+    <Reference Include="Newtonsoft.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed, processorArchitecture=MSIL">
+      <HintPath>..\..\packages\Newtonsoft.Json.6.0.4\lib\net45\Newtonsoft.Json.dll</HintPath>
+    </Reference>
```
Comment (vit-svoboda) [inline TodoList/sources/TodoList.Contracts/TodoList.Contracts.csproj]: Do you really need these references in the contracts project?

File: TodoList/sources/TodoList.Repository/packages.config [hunk 15]
```diff
@@ -1,5 +1,9 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
   <package id="CommonServiceLocator" version="1.3" targetFramework="net46" />
+  <package id="MongoDB.Bson" version="2.4.4" targetFramework="net46" />
+  <package id="MongoDB.Driver" version="2.4.4" targetFramework="net46" />
+  <package id="MongoDB.Driver.Core" version="2.4.4" targetFramework="net46" />
+  <package id="System.Runtime.InteropServices.RuntimeInformation" version="4.0.0" targetFramework="net46" />
```
Comment (vit-svoboda) [inline TodoList/sources/TodoList.Repository/packages.config]: That's a surprising choice of a NuGet. Are you doing something with OS runtime in your repository?

File: TodoList/sources/TodoList.Repository/Dependency/Bootstrapper.cs [hunk 16]
```diff
@@ -4,7 +4,7 @@
 
 namespace TodoList.Repository.Dependency
 {
-    public class RegisterTypes : IBootstrapper
+    public class Bootstrapper : IBootstrapper
     {
         public IUnityContainer RegisterType(IUnityContainer container) 
             => container.RegisterType<INodesRepository, NodesRepository>(new HierarchicalLifetimeManager());
```
Comment (vit-svoboda) [inline TodoList/sources/TodoList.Repository/Dependency/Bootstrapper.cs]: I believe Petr was hinting to use a different lifetime manager.

File: TodoList/sources/TodoList.Services/Dependency/Bootstrapper.cs [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+using TodoList.Services.Static_Wrappers;
+using TodoList.Services.Nodes;
+
+namespace TodoList.Services.Dependency
+{
+    public class Bootstrapper : IBootstrapper
+    {
+        public IUnityContainer RegisterType(IUnityContainer container)
+        {
+            container.RegisterType<IGenerateIdService, GenerateIdService>(new HierarchicalLifetimeManager());
```
Comment (vit-svoboda) [inline TodoList/sources/TodoList.Services/Dependency/Bootstrapper.cs]: These stateless services could be singleton or transient, but scoped is an odd choice.

File: TodoList/sources/TodoList.Services/Nodes/CreateNodeService.cs [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+        {
+            _repository = repository;
+            _generateIdService = generateIdService;
+            _timeService = timeService;
+        }
+
+        public async Task<NodeModel> CreateNodeAsync(NodeModel node)
+            => await CreateNodeAsync(node, _generateIdService.GenerateId());
+
+        private async Task<NodeModel> CreateNodeAsync(NodeModel node, Guid id)
```
Comment (vit-svoboda) [inline TodoList/sources/TodoList.Services/Nodes/CreateNodeService.cs]: I can't see the point of this function. Is there a reason for the 2 CreateNodeAsync functions?

File: TodoList/tests/TodoList.Services.Tests/CreateNodeServiceTests.cs [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+
+            var actualNode = await _createNodeService.CreateNodeAsync(testModel);
+
+            Assert.That(actualNode, Is.EqualTo(expectedNode).UsingNodeModelEqualityComparer());
+        }
+
+        [Test]
+        public async Task CreateNodeAsync_WithNullData_ThrowsException()
+        {
+            try
```
Comment (vit-svoboda) [inline TodoList/tests/TodoList.Services.Tests/CreateNodeServiceTests.cs]: Never try in tests. If you expect an exception to be thrown, use `Assert.Throws(lambda)` or `Assert.That(lambda Throws)`.

File: TodoList/tests/TodoList.Services.Tests/UpdateNodeServiceTests.cs [hunk 20]
```diff
@@ -0,0 +1,112 @@
+﻿using System;
+using System.Threading.Tasks;
+using NSubstitute;
+using NUnit.Framework;
+using TodoList.Api.Tests.Extensions;
+using TodoList.Contracts.Models;
+using TodoList.Contracts.Repository;
+using TodoList.Contracts.Services;
+using TodoList.Services.Nodes;
+using Assert = NUnit.Framework.Assert;
```
Comment (vit-svoboda) [inline TodoList/tests/TodoList.Services.Tests/UpdateNodeServiceTests.cs]: You don't need this.

**Review by vit-svoboda — CHANGES_REQUESTED** [review]
A nice solution indeed. Just could use a little bit of cleanup. Except for the try-catch in the tests, that's naughty!

File: TodoList/sources/TodoList.Api/Controllers/NodesController.cs (hunk 13 above)
Comment (PetrSvirak) [reply TodoList/sources/TodoList.Api/Controllers/NodesController.cs]: Sorry to intervene, but while it is completely correct suggestion to create a view model and it would actually be a correct solution, for the educational purses (~ to see how it sucks, both in API and connected front-end): please, validate in the controller that the `id` in model is either same as the one in URL or (~xor) require the `id` to be empty (as it has to be in `POST`)

File: TodoList/sources/TodoList.Api/Controllers/NodesController.cs [hunk 21]
```diff
@@ -110,7 +116,11 @@ private void ValidatePutNodeModel(NodeModel node)
             }
             if (node.Id == Guid.Empty)
```
Comment (vit-svoboda) [inline TodoList/sources/TodoList.Api/Controllers/NodesController.cs]: I'd say from the usability point of the API, a missing id in the body is fine. You just put `{ "text": theText }` to the correct URL. However, if you insist on the id in PUT body, why not ;)

File: TodoList/tests/TodoList.Services.Tests/CreateNodeServiceTests.cs [hunk 22]
```diff
@@ -53,23 +52,18 @@ public void SetUp()
...
-                const string expectedMessage = "Text to add has to be provided.";
+            var expectedMessage = "Text to add has to be provided.";
 
-                Assert.That(expectedMessage, Is.EqualTo(exception.Message));
-            }
-            catch (Exception exception)
-            {
-                Assert.Fail("A different type of exception has been thrown: " + exception.GetType());
-            }
+            async Task<NodeModel> CreateNullNodeAsync() => await _createNodeService.CreateNodeAsync(null);
```
Comment (vit-svoboda) [inline TodoList/tests/TodoList.Services.Tests/CreateNodeServiceTests.cs]: I would inline the lambda in the `Assert.That` call. To be honest, I thought you're trying to hack this way around the fact `Assert.That` doesn't support async delegates. But to my surprise it does. In earlier versions of NUnit, you'd have to use `await Assert.ThrowsAsync`.

**Review by vit-svoboda — CHANGES_REQUESTED** [review]
Nicely done. Just the change in tests that expect an exception to be thrown kind of confused me.

File: TodoList/sources/TodoList.Api/Controllers/NodesController.cs [hunk 23]
```diff
@@ -114,13 +114,9 @@ private void ValidatePutNodeModel(Guid id, NodeModel node)
                 ModelState.AddModelError(nameof(node), "Node model is not correctly defined.");
                 return;
             }
-            if (node.Id == Guid.Empty)
-            {
-                ModelState.AddModelError(nameof(node.Id), "Node model must have id parameter specified here.");
-            }
-            if (id != node.Id)
+            if (node.Id != Guid.Empty)
```
Comment (vit-svoboda) [inline TodoList/sources/TodoList.Api/Controllers/NodesController.cs]: It's interesting you mostly use `default(Guid)`, but here you use `Guid.Empty`. Makes no difference, though ;)

**Review by vit-svoboda — APPROVED** [review]
Looks good. This seems to be all of the back-end for you ;)
